vshard/router/init.lua

```diff
@@ -796,6 +799,28 @@ local function router_cfg(router, cfg, is_reload)
     end
 end
 
+--
+-- Synchronize update of router configuration from mutiple fibers.
+-- At each point of time only single fiber can execute 'router_cfg' for particular router.
+-- Other fibers are waiting for current call of 'router_cfg' to return.
+--
+local function apply_router_cfg(router, cfg, is_reload)
+    if not router.cfg_lock:put(true, consts.ROUTER_CFG_TIMEOUT) then
+        error('Configuration timeout')
+    end
+
+    if M.errinj.ERRINJ_CONCURRENT_CFG_TIMEOUT then
+        fiber.sleep(consts.ROUTER_CFG_TIMEOUT * 2)
```

Comment:
I would not use any exact timings here. When you rely on timeouts, you may create a 'flaky' test, which will sometimes fail, and is usually hard to catch and fix. We already have plenty of them. So timeouts should be avoided when possible. Also you used really high timeout here - 2 seconds just for 1 test case. Taking into account, that parallel tests does not work here, it is a significant slowdown.

Instead, I propose you to do this.
1) Rename ERRINJ_CONCURRENT_CFG_TIMEOUT to ERRINJ_CONCURRENT_CFG_COUNT.
2) In a test set it to 0.
3) Start 2 fibers. Each time a fiber goes here and sees ERRINJ_CONCURRENT_CFG_DELAY ~= false, it increments it, and makes assertion that the counter == 1. And decrements back in the end of apply_router_cfg.